src/southbridge/amd/sb700/sm.c

```diff
@@ -0,0 +1,517 @@
+		byte |= 1 << 0;
+	}
+	/*Set bit2 to 1, enable Io port 60h read/write SMi trapping and
+	 *Io port 64h write Smi trapping. conflict with ps2 keyboard
+	 */
+	//byte |= 1 << 2 | 1 << 3 | 1 << 4;
+	byte |= 1 << 3 | 1 << 4;
+	pci_write_config8(dev, 0xE1, byte);
+
+	/* 2.5 Enabling Non-Posted Memory Write */
+	axindxc_reg(0x10, 1 << 9, 1 << 9);
+
+	/* 2.11 IO Trap Settings */
+	abcfg_reg(0x10090, 1 << 16, 1 << 16);
+
+	/* ab index */
+	pci_write_config32(dev, 0xF0, AB_INDX);
+	/* Initialize the real time clock */
+	cmos_init(0);
+
+	/* 4.3 Enabling Upstream DMA Access */
+	axcfg_reg(0x04, 1 << 2, 1 << 2);
+	/* 4.4 Enabling IDE/PCIB Prefetch for Performance Enhancement */
+	abcfg_reg(0x10060, 9 << 17, 9 << 17);
+	abcfg_reg(0x10064, 9 << 17, 9 << 17);
+
+	/* 4.5 Enabling OHCI Prefetch for Performance Enhancement, A12 */
+	abcfg_reg(0x80, 1 << 0, 1<< 0);
+
+	/* 4.6 B-Link Client's Credit Variable Settings for the Downstream Arbitration Equation */
+	/* 4.7 Enabling Additional Address Bits Checking in Downstream */
+	/* 4.16 IO write and SMI ordering enhancement*/
+	abcfg_reg(0x9c, 3 << 0, 3 << 0);
+	if (REV_SB700_A12 == rev) {
+		abcfg_reg(0x9c, 1 << 8, 1 << 8);
+	} else if (rev >= REV_SB700_A14) {
+		abcfg_reg(0x9c, 1 << 8, 0 << 8);
+	}
+	if (REV_SB700_A15 == rev) {
+		abcfg_reg(0x90, 1 << 21, 1 << 21);
+		abcfg_reg(0x9c, 1 << 5 | 1 << 9 | 1 << 15, 1 << 5 | 1 << 9 | 1 << 15);
+	}
+
+	/* 4.8 Set B-Link Prefetch Mode */
+	abcfg_reg(0x80, 3 << 17, 3 << 17);
+
+	/* 4.9 Enabling Detection of Upstream Interrupts */
+	abcfg_reg(0x94, 1 << 20 | 0x7FFFF, 1 << 20 | 0x00FEE);
+
+	/* 4.10: Enabling Downstream Posted Transactions to Pass Non-Posted
+	 *  Transactions for the K8 Platform (for All Revisions) */
+	abcfg_reg(0x10090, 1 << 8, 1 << 8);
+
+	/* Set ACPI Software clock Throttling Period to 244 us*/
+	byte = pm_ioread(0x68);
+	byte &= ~(3 << 6);
+	byte |= (2 << 6);	/* 244us */
+	pm_iowrite(0x68, byte);
+
+	if (REV_SB700_A15 == rev) {
+		u16 word;
+
+		/* rpr v2.13 4.18 Enabling Posted Pass Non-Posted Downstream */
+		axindxc_reg(0x02, 1 << 9, 1 << 9);
+		abcfg_reg(0x9C, 0x00007CC0, 0x00007CC0);
+		abcfg_reg(0x1009C, 0x00000030, 0x00000030);
+		abcfg_reg(0x10090, 0x00001E00, 0x00001E00);
+
+		/* rpr v2.13 4.19 Enabling Posted Pass Non-Posted Upstream */
+		abcfg_reg(0x58, 0x0000F800, 0x0000E800);
+
+		/* rpr v2.13 4.20 64 bit Non-Posted Memory Write Support */
+		axindxc_reg(0x02, 1 << 10, 1 << 10);
+
+		/* rpr v2.13 2.38 Unconditional Shutdown */
+		byte = pci_read_config8(dev, 0x43);
+		byte &= ~(1 << 3);
+		pci_write_config8(dev, 0x43, byte);
+
+		word = pci_read_config16(dev, 0x38);
+		word |= 1 << 12;
+		pci_write_config16(dev, 0x38, word);
+
+		byte |= 1 << 3;
+		pci_write_config8(dev, 0x43, byte);
+
+		/* Enable southbridge MMIO decode */
+		dword = pci_read_config32(dev, SB_MMIO_CFG_REG);
+		dword &= ~(0xffffff << 8);
+		dword |= SB_MMIO_BASE_ADDRESS;
+		dword |= 0x1;
+		pci_write_config32(dev, SB_MMIO_CFG_REG, dword);
```

Comment:
This should be done in set_resources(). Right now set_resources() enables decoding without setting base address, which results in lack of serial output between those two points.

Note that I haven't checked if anything enabled by this register is actually used during that period and if it's safe to move it. Then again, accessing something under BAR before it is set wouldn't be good either.